pegasus/test/test_dev_routes.rb

```diff
@@ -0,0 +1,154 @@
+require 'minitest/autorun'
+require 'rack/test'
+require 'mocha/mini_test'
+require_relative 'fixtures/mock_pegasus'
+
+BUILD_STARTED_PATH = deploy_dir('build-started')
+FAKE_START_BUILD_TOKEN = 'fake-start-build-token'
+
+class DevRoutesTest < Minitest::Test
+  describe '/api/dev/start-build' do
+    before do
+      $log.level = Logger::ERROR # Pegasus spams debug logging otherwise
+
+      # Fake slack token for tests
+      CDO.slack_start_build_token = FAKE_START_BUILD_TOKEN
+
+      # Should have no build-started file at the beginning of tests
+      system "rm -f #{BUILD_STARTED_PATH}"
+      refute File.file?(BUILD_STARTED_PATH), 'Precondition failure: File start-build should not exist.'
+    end
+
+    it 'is forbidden on production environments' do
+      assert_forbidden_on_rack_env :production
+    end
+
+    it 'is forbidden on development environments' do
+      assert_forbidden_on_rack_env :development
+    end
+
+    def assert_forbidden_on_rack_env(forbidden_env)
+      in_rack_env(forbidden_env) do
+        pegasus = make_test_pegasus
+
+        pegasus.post('/api/dev/start-build', {
+          token: FAKE_START_BUILD_TOKEN,
+          user_name: 'Dave'
+        })
+        assert_equal 403, pegasus.last_response.status
+        refute File.file? BUILD_STARTED_PATH
+      end
+    end
+
+    it 'is allowed on staging' do
+      assert_allowed_on_rack_env(:staging)
+    end
+
+    it 'is allowed on test' do
+      assert_allowed_on_rack_env(:test)
+    end
+
+    it 'is allowed on adhoc' do
+      assert_allowed_on_rack_env(:adhoc)
+    end
+
+    it 'is allowed on levelbuilder' do
+      assert_allowed_on_rack_env(:levelbuilder)
+    end
+
+    def assert_allowed_on_rack_env(allowed_env)
+      in_rack_env(allowed_env) do
+        pegasus = make_test_pegasus
+        begin
+          pegasus.post('/api/dev/start-build', {
+              token: FAKE_START_BUILD_TOKEN,
+              user_name: 'Dave'
+          })
+          assert_equal 200, pegasus.last_response.status
+        ensure
+          system "rm -f #{BUILD_STARTED_PATH}"
+        end
+      end
+    end
+
+    it 'is forbidden with a missing token' do
+      in_rack_env(:staging) do
+        pegasus = make_test_pegasus
+        pegasus.post('/api/dev/start-build', {
+            user_name: 'Dave'
+        })
+        assert_equal 403, pegasus.last_response.status
+      end
+    end
+
+    it 'is forbidden with an incorrect token' do
+      in_rack_env(:staging) do
+        pegasus = make_test_pegasus
+        pegasus.post('/api/dev/start-build', {
+            token: 'incorrect-token',
+            user_name: 'Dave'
+        })
+        assert_equal 403, pegasus.last_response.status
+      end
+    end
+
+    it 'generates a start_build file if none exists' do
+      in_rack_env(:test) do
+        pegasus = make_test_pegasus
+        begin
+          refute File.file? BUILD_STARTED_PATH
```

Comment:
Intentional - I felt like it communicated the before/after assertion of this particular test case better.